specs/dynamicdependencies/DynamicDependencies.md

```diff
@@ -0,0 +1,1409 @@
+    /// and this option is true the package is prepended to the set of packages of the same rank.
+    /// By default packages are appended to the set of packages with the same rank.
+    Boolean PrependIfRankCollision;
+}
+
+/// TBD
+runtimeclass PackageDependency
+{
+    /// Create an intstance of the package dependency identified by id.
+    PackageDependency(String id);
+
+    /// Return the package dependency id.
+    String Id { get; }
+
+    /// Return true if the package dependencies would produce the same
+    /// package when resolved e.g. whether they share the same packageFamilyName,
+    /// minVersion, and packageDependencyProcessorArchitectures values.
+    static Boolean AreEquivalent(String packageDependencyId1, String packageDependencyId2);
+
+    /// Return true if this and other would produce the same
+    /// package when resolved e.g. whether they share the same packageFamilyName,
+    /// minVersion, and packageDependencyProcessorArchitectures values.
+    Boolean IsEquivalent(PackageDependency other);
+
+    /// Define a package dependency for the current user. The criteria for a PackageDependency
+    /// (package family name, minimum version, etc) may match multiple
+    /// packages, but ensures Deployment won't remove a package if it's
+    /// the only one satisfying the PackageDependency.
+    ///
+    /// @note A package matching a PackageDependency pin can still be removed
+    ///       as long as there's another that satisfies the PackageDependency.
+    ///       For example, if Fwk-v1 is installed and a PackageDependency specifies
+    ///       MinVersion=1 and then Fwk-v2 is installed, Deployment could remove
+    ///       Fwk-v1 because Fwk-v2 will satisfy the PackageDependency. After Fwk-v1
+    ///       is removed Deployment won't remove Fwk-v2 because it's the only package
+    ///       satisfying the PackageDependency. Thus  Fwk-v1 and Fwk-v2 (and any other
+    ///       package matching the PackageDependency) are 'loosely pinned' â€“ Deployment
+    ///       guarantees it won't remove a package if it would make a PackageDependency
+    ///       unsatisfied.
+    ///
+    /// A PackageDependency specifies criteria (package family, minimum version, etc)
+    /// and not a specific package. Deployment reserves the right to use a different
+    /// package (e.g. higher version) to satisfy the PackageDependency if/when
+    /// one becomes available.
+    ///
+    /// This method is equivalent to PinForUser(null,...).
+    ///
+    /// @param packageFamilyName the package family to pin
+    /// @param minVerrsion the minimum version to pin
+    ///
+    /// @note This fails if the package dependency cannot be resolved to a specific package.
+    ///
+    /// @see Pin(String, PackageVersion, PinPackageDependencyOptions)
+    /// @see PinForUser()
+    /// @see PinForSystem()
+    static PackageDependency Pin(
```

Comment:
The fact that you create a PackageDependency by calling `Pin` highlights the weirdness of the terminology. Shouldn't the method for creating a PackageDependency be called "Create"? People are not going to browse the IntelliSense and say "Oh, yeah, Pin, that's the method I want to use to get the party started."

Reply:
RECOMMEND: Use `TryCreate` here for creating, and `Delete` as the opposite verb.

CONSIDER: Add a sequence diagram showing lifecycle of create, store, activate, delete.

Other names Create/Delete, Register/Deregister, Add/Remove, Addref/Release.  "Register" here is a Word of Power around packages.

DISCUSSION: Why do we have both flat-C and WinRT types in this review? To try out a few things and because this is part of app bootstrapping where we can't depend on the app having access to FWPs first to use these and get the WinRT metadata support.